### PR TITLE
4594 - fix unnecessary-constructor parameters for super call

### LINKED ISSUE #4594: unnecessary-constructor parameters for super call
### Rule Suggestion
So there is already an unnecessary-constructor but it does not check for this construction:
```typescript
constructor(requiredParam: string) {
    super(requiredParam);
}
```
While this is also unnecessary and it should be great if this would also trigger the unnecessary-constructor rule.

I wasn't sure in which category I should have posted this. This is isnt realy a "new rule" but wasn't a "bug" either.

### PR DESCRIPTION
**Pull request by tanmoyopenroot**
#### PR checklist

- [x] Addresses an existing issue: fixes #4594 
- [x] enhancement
- [x] Includes tests
- [ ] Documentation update

#### Overview of change:
Add `check-super-calls` option to `unnecessary-constructor`
```javascipt
{
  "rules": {
    "unnecessary-constructor": [
        true,
        { "check-super-calls": true }
    ]
  }
}
```
#### Is there anything you'd like reviewers to focus on?
Not sure if the option name `check-super-calls` is correct.

#### CHANGELOG.md entry:

[new-rule-option] `check-super-calls` option for `unnecessary-constructor` rule

### CODE REVIEW

File: src/rules/unnecessaryConstructorRule.ts [hunk 1]
```diff
@@ -15,18 +15,36 @@
...
 import * as ts from "typescript";
 
 import * as Lint from "../index";
 import { Replacement } from "../language/rule/rule";
 
+interface Options {
+    unnecessarySuperCall: boolean;
+}
+
+const OPTION_UNNECESSARY_SUPER_CALL = "unnecessary-super-call";
```
Comment (JoshuaKGoldberg) [inline src/rules/unnecessaryConstructorRule.ts]: Yeah, I'm not sold on this option name either... "Unnecessary super call" implies the `super` is the unnecessary thing, not the whole `constructor`. How about... `check-super-calls`?

File: src/rules/unnecessaryConstructorRule.ts [hunk 2]
```diff
@@ -39,12 +57,49 @@ export class Rule extends Lint.Rules.AbstractRule {
...
+    }
+
+    return false;
+};
+
+const isEmptyOrContainsOnlySuper = (node: ts.ConstructorDeclaration, options: Options): boolean => {
+    if (node.body) {
+        const { unnecessarySuperCall } = options;
+
+        if (!containsStatements(node.body.statements)) {
```
Comment (JoshuaKGoldberg) [inline src/rules/unnecessaryConstructorRule.ts]: Seems like it'd be a bit less code to just say `if (node.body.statements.length === 0)`, no? Let's change to that.

File: src/rules/unnecessaryConstructorRule.ts [hunk 3]
```diff
@@ -39,12 +57,49 @@ export class Rule extends Lint.Rules.AbstractRule {
...
+        return this.applyWithFunction(sourceFile, walk, options);
     }
 }
 
-const isEmptyConstructor = (node: ts.ConstructorDeclaration): boolean =>
-    node.body !== undefined && node.body.statements.length === 0;
+const containsStatements = (statements: ts.NodeArray<ts.Statement>) => statements.length > 0;
+
+const containsSuper = (statements: ts.NodeArray<ts.Statement>): boolean => {
+    for (const statement of statements) {
```
Comment (JoshuaKGoldberg) [inline src/rules/unnecessaryConstructorRule.ts]: Here, you loop over all statements, but `containsSuper` is only called if `statements` is length 1. You can just have this take in a single `statement: ts.Statement`.

**Comment by JoshuaKGoldberg** [comment]
Looks like you have some [lint failures](https://circleci.com/gh/palantir/tslint/17033?utm_campaign=vcs-integration-link&utm_medium=referral&utm_source=github-build-link) you'll need to clean up too:

```
This type is not allowed in the 'if' condition because it could be undefined. Allowed types are boolean or boolean-or-undefined. (strict-boolean-expressions)
  87 | 
  88 | const isEmptyOrContainsOnlySuper = (node: ts.ConstructorDeclaration, options: Options): boolean => {
> 89 |     if (node.body) {
     |        ^
  90 |         const { unnecessarySuperCall } = options;
  91 | 
  92 |         if (!containsStatements(node.body.statements)) {
```

File: test/rules/unnecessary-constructor/unnecessary-super-call/test.ts.lint [hunk 4]
```diff
@@ -0,0 +1,137 @@
...
+}
+
+class X {
+    constructor(private param1: number, param2: number) {}
+}
+
+export class Y extends X {
+    constructor(param1: number) {
+    ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+        super(param1, 2);
```
[the remark this answers is not in the document]
Comment (tanmoyopenroot) [reply test/rules/unnecessary-constructor/unnecessary-super-call/test.ts.lint]: Done

File: src/rules/unnecessaryConstructorRule.ts (hunk 1 above)
Comment (tanmoyopenroot) [reply src/rules/unnecessaryConstructorRule.ts]: Done

File: src/rules/unnecessaryConstructorRule.ts (hunk 2 above)
Comment (tanmoyopenroot) [reply src/rules/unnecessaryConstructorRule.ts]: Done

File: src/rules/unnecessaryConstructorRule.ts (hunk 3 above)
Comment (tanmoyopenroot) [reply src/rules/unnecessaryConstructorRule.ts]: Done

**Comment by tanmoyopenroot** [comment]
> Looks like you have some [lint failures](https://circleci.com/gh/palantir/tslint/17033?utm_campaign=vcs-integration-link&utm_medium=referral&utm_source=github-build-link) you'll need to clean up too:
> 
> ```
> This type is not allowed in the 'if' condition because it could be undefined. Allowed types are boolean or boolean-or-undefined. (strict-boolean-expressions)
>   87 | 
>   88 | const isEmptyOrContainsOnlySuper = (node: ts.ConstructorDeclaration, options: Options): boolean => {
> > 89 |     if (node.body) {
>      |        ^
>   90 |         const { unnecessarySuperCall } = options;
>   91 | 
>   92 |         if (!containsStatements(node.body.statements)) {
> ```

Done

File: test/rules/unnecessary-constructor/unnecessary-super-call/test.ts.lint [hunk 5]
```diff
@@ -0,0 +1,137 @@
...
+}
+
+class Super extends Base {
+    public param1: number;
+    constructor(public param2: number) {
+        super(param2);
+        this.param1 = param2;
+    }
+}
+
```
[the remark this answers is not in the document]
Comment (tanmoyopenroot) [reply test/rules/unnecessary-constructor/unnecessary-super-call/test.ts.lint]: Done

**Comment by tanmoyopenroot** [comment]
> A good start! 
> Left some comments on requested structural changes and missing test changes.

Done

**Comment by tanmoyopenroot** [comment]
@JoshuaKGoldberg  @adidahiya ???

**Comment by adidahiya** [comment]
testNext failure is unreleated. I'd like to release this and improve it in future PRs if necessary

**Pull request merged**